### PR TITLE
feat(repository): add more helpers for HasManyThrough

### PR DESCRIPTION
**Pull request by agnes512**
Add two more helpers for HasManyThrough relation

## Checklist

👉 [Read and sign the CLA (Contributor License Agreement)](https://cla.strongloop.com/agreements/strongloop/loopback-next) 👈

- [x] `npm test` passes on your machine
- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- [x] API Documentation in code was updated
- [ ] Documentation in [/docs/site](../tree/master/docs/site) was updated
- [ ] Affected artifact templates in `packages/cli` were updated
- [ ] Affected example projects in `examples/*` were updated

👉 [Check out how to submit a PR](https://loopback.io/doc/en/lb4/submitting_a_pr.html) 👈

### CODE REVIEW

File: packages/repository/src/relations/has-many/has-many-through.helpers.ts [hunk 1]
```diff
@@ -109,6 +117,45 @@ export function createThroughConstraint<Through extends Entity, ForeignKeyType>(
...
+ * };
+ * createThroughConstraint(resolvedMetadata, {id: 3, name: 'a product'});
+ *
+ * >>> {'productId': 1}
+ *
+ *  * createThroughConstraint(resolvedMetadata, {id: {inq:[3,4]}});
+ *
+ * >>> {'productId': {inq:[3,4]}}
+ */
+export function createThroughFkConstraint<Target, Through extends Entity>(
```
Comment (agnes512) [inline packages/repository/src/relations/has-many/has-many-through.helpers.ts:147]: This new helper is created for deleting through models once the target instances get deleted

**Comment by agnes512** [comment]
Added two helpers. They are all for deleting through models as we decided to have the delete method this way https://github.com/strongloop/loopback-next/pull/4438/files#r402105078

File: packages/repository/src/__tests__/unit/repositories/constraint-utils.unit.ts [hunk 2]
```diff
@@ -77,6 +78,23 @@ describe('constraint utility functions', () => {
...
+    };
+    it('enforces a constraint', () => {
+      const constraint = [{id: '5'}, {y: 'y'}];
+      const result = constrainWhereOr(inputWhere, constraint);
+      expect(result).to.deepEqual({...inputWhere, or: constraint});
+    });
+
+    it('enforces constraint with dup key', () => {
+      const constraint = [{y: 'z'}];
+      const result = constrainWhereOr(inputWhere, constraint);
```
Comment (bajtos) [inline packages/repository/src/__tests__/unit/repositories/constraint-utils.unit.ts:93]: I am little bit confused. The test name says "with dup key", but IIUC, the code is using keys `x` and `y`. Where is the duplicated key?

Comment (agnes512) [reply packages/repository/src/__tests__/unit/repositories/constraint-utils.unit.ts:93]: @bajtos My bad, forgot to add the key back. Fixed.

File: packages/repository/src/relations/has-many/has-many-through.helpers.ts [hunk 3]
```diff
@@ -109,6 +119,45 @@ export function createThroughConstraint<Through extends Entity, ForeignKeyType>(
...
+ *  keyTo: 'id',
+ *  through: {
+ *    model: () => CategoryProductLink,
+ *    keyFrom: 'categoryId',
+ *    keyTo: 'productId',
+ *  },
+ * };
+ * createThroughConstraint(resolvedMetadata, {id: 3, name: 'a product'});
+ *
+ * >>> {'productId': 1}
```
Comment (jannyHou) [inline packages/repository/src/relations/has-many/has-many-through.helpers.ts:143]: IIUC the constraint here should be `{productId: 1}`?

File: packages/repository/src/relations/has-many/has-many-through.helpers.ts [hunk 4]
```diff
@@ -109,6 +119,45 @@ export function createThroughConstraint<Through extends Entity, ForeignKeyType>(
...
+ *  // .. other props
+ *  keyFrom: 'id',
+ *  keyTo: 'id',
+ *  through: {
+ *    model: () => CategoryProductLink,
+ *    keyFrom: 'categoryId',
+ *    keyTo: 'productId',
+ *  },
+ * };
+ * createThroughConstraint(resolvedMetadata, {id: 3, name: 'a product'});
```
Comment (jannyHou) [inline packages/repository/src/relations/has-many/has-many-through.helpers.ts:141]: The additional query field reminds me sth(sorry don't remember where is the original chat :p) : 
 
We want to distinguish the query on through model(CategoryProductLink) and target model(Product), how do we achieve that? 

For example, `{name: 'a product'}` seems for target model, and `createThroughConstraint` applies for through model, and that's why `name` is ignored here. But:

- where is the place that processes the target model's filter?
- how to specify the through model's filter?

(I started from the helper function, might missing the code/docs if they are created other places.)

Comment (agnes512) [reply packages/repository/src/relations/has-many/has-many-through.helpers.ts:141]: @jannyHou I think these helpers are just for handling the relation metadata. And filters are handled in the hasManyThrough repository. Here is an example in the original design: [find method](https://github.com/strongloop/loopback-next/pull/2359/files#diff-bd2fafeb3bd8f873174e06856a1d7471R124-R141). They won't be included in this PR. But I'd love to discuss it: 

So far from those two community PRs(#2359 , #4438 ), I only see the target  model is capable of applying filters. Not sure if we need it for through models ( I don't see the reason to allow filtering through models tho). Do we need it?

**Pull request merged**